linkerd/linkerd2
# Add routes command

Add a routes command which displays per-route stats for services that have service profiles defined.

```
$ bin/linkerd routes webapp --from deploy/traffic
               ROUTE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TLS
/authors/{id}/delete     5.64%   7.4rps          37ms         247ms         695ms    0%
  /books/{id}/delete     5.75%   7.2rps          29ms         160ms         680ms    0%
              /books    44.64%   1.9rps          71ms        1300ms        1860ms    0%
         /books/{id}   100.00%   0.8rps          75ms         470ms         494ms    0%
     /book/{id}/edit    50.00%   0.8rps         138ms         265ms         293ms    0%
       /authors/{id}   100.00%   0.4rps          30ms          95ms          99ms    0%
            /authors   100.00%   0.4rps          30ms          90ms          98ms    0%
```

This change has three parts:
* A new public-api RPC called `TopRoutes` which serves per-route stat data about a service
* An implementation of TopRoutes in the public-api service.  This implementation reads per-route data from Prometheus.  This is very similar to how the StatSummaries RPC and much of the code was able to be refactored and shared.
* A new CLI command called `routes` which displays the per-route data in a tabular or json format.  This is very similar to the `stat` command and much of the code was able to be refactored and shared.

The easiest way to test this change is to do the following:
1. install Linkerd off this branch
1. inject and deploy the books app (https://run.linkerd.io/booksapp.yml)
1. apply this service profile: https://gist.github.com/adleong/1f2d1d79f93f621cdd64f852f720fa16
1. run `bin/linkerd routes webapp --from deploy/traffic`

Note that as of the currently targeted proxy version, only outbound route stats are supported so the `--from` flag must be included in order to see data.  This restriction will be lifted in an upcoming PR once we add support for inbound route stats as well.
